### ISSUE TITLE
Doc for how to do cookie auth?

### ISSUE THREAD

**Issue by samsp-msft** [microsoft/reverse-proxy#990]
How to handle multiple sites with different auth requirements, How to make sure each site has its own auth cookies and they don't overwrite each other

**Comment by Tratcher** [comment]
YARP isn't directly involved in any of this, there are no YARP settings for cookies.

General guidance like the following would apply to the application:
https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/web-farm?view=aspnetcore-5.0
https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-5.0

It would also depend on how they set up their routes. If the routes are host based then there's no conflict. If the apps are path based then either the cookies need to also include a path, or the apps need to use unique cookie names.

It's also possible for destination apps to share cookies using shared state like encryption keys.

**Comment by samsp-msft** [comment]
I am thinking about the case where the proxy is doing URL munging so that the URL that the client sees is different from the back-end servers.
If both apps are using the same host name for the front end, then their cookies will all be under the same domain, and so have a high chance of colliding with each other. Both hosts must take care not to overwrite each other, or worse throw away the values that they don't understand. This may be an app config problem rather than YARP.

If the proxy is changing the domain name while addressing the backend(s) then will the cookies will need to be re-written to have the correct domain names when the client sees them, or is that implicit from the domain of the server?

**Comment by Tratcher** [comment]
> If YARP is doing the cookie auth itself, it may need to treat each route as a different context for authentication and authorization.

AuthZ is configured per route, but it is up to the user to work out polices that make sense for their architecture.

> If the proxy is changing the domain name while addressing the backend(s) then will the cookies will need to be re-written to have the correct domain names when the client sees them, or is that implicit from the domain of the server?

https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-5.0 covers doing fixup in the destination app so that it can generate the proper links, cookies, etc. using public values. Re-writing cookie attributes in the proxy is discouraged but possible using response header transforms.

In general though, while the set-cookie header does have a domain field, it's common for people to let the client set an implicit default.

**Comment by karelz** [comment]
Triage: Should be covered by other docs - @Tratcher can you please confirm?

**Comment by Tratcher** [comment]
Auth per rout in the proxy is covered by https://github.com/microsoft/reverse-proxy/blob/main/docs/docfx/articles/authn-authz.md

Set-Cookie header issues are covered by https://github.com/microsoft/reverse-proxy/blob/main/docs/docfx/articles/header-guidelines.md#set-cookie

**Comment by MartyBJones** [comment]
I figured out the issue.  I was initializing the response transformer as below.  I was thinking the "AddPathPrefix" was to tell the context to only apply for that the transform to a given path but it was actually doing an request transform which is not what I wanted.  I commented out that and it works as expected now.
`proxyBuilder.AddTransforms(transformBuilderContext => {
    //transformBuilderContext.AddPathPrefix("/legacy");
    transformBuilderContext.ResponseTransforms.Add(new CustomLocationTransformer());
});
`
